notebooks/2021_final_submission/ncaa-march-madness-2021.ipynb

```diff
@@ -0,0 +1,1273 @@
+    "sub[\"Seed_diff\"] = sub[\"T1_seed\"] - sub[\"T2_seed\"]\n",
+    "sub.head()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "len(sub)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "_uuid": "cb1627abe53f9db9fdb5be2e45a739e0f78ede83"
+   },
+   "outputs": [],
+   "source": [
+    "Xsub = sub[features].values\n",
+    "dtest = xgb.DMatrix(Xsub)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "_uuid": "a700a993c601d9692f2374f11c8eadb234979086",
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "sub_models = []\n",
+    "for i in range(repeat_cv):\n",
+    "    print(f\"Fold repeater {i}\")\n",
+    "    sub_models.append(\n",
+    "        xgb.train(\n",
+    "          params = param,\n",
+    "          dtrain = dtrain,\n",
+    "          num_boost_round = int(iteration_counts[i] * 1.05),\n",
+    "          verbose_eval = 50\n",
+    "        )\n",
+    "    )"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "_uuid": "01757a38a0ad222dec31bf062ceee7cfb61a6b16"
+   },
+   "outputs": [],
+   "source": [
+    "sub_preds = []\n",
+    "for i in range(repeat_cv):\n",
+    "    sub_preds.append(np.clip(spline_model[i](np.clip(sub_models[i].predict(dtest),-30,30)),0.025,0.975))\n",
+    "    \n",
+    "sub[\"Pred\"] = pd.DataFrame(sub_preds).mean(axis=0)\n",
+    "\n",
+    "sub.loc[(sub.T1_seed==1) & (sub.T2_seed==16), 'Pred'] = 1.0\n",
+    "sub.loc[(sub.T1_seed==2) & (sub.T2_seed==15), 'Pred'] = 1.0\n",
+    "sub.loc[(sub.T1_seed==3) & (sub.T2_seed==14), 'Pred'] = 1.0\n",
+    "sub.loc[(sub.T1_seed==4) & (sub.T2_seed==13), 'Pred'] = 1.0\n",
+    "sub.loc[(sub.T1_seed==16) & (sub.T2_seed==1), 'Pred'] = 0.0\n",
+    "sub.loc[(sub.T1_seed==15) & (sub.T2_seed==2), 'Pred'] = 0.0\n",
+    "sub.loc[(sub.T1_seed==14) & (sub.T2_seed==3), 'Pred'] = 0.0\n",
+    "sub.loc[(sub.T1_seed==13) & (sub.T2_seed==4), 'Pred'] = 0.0\n",
+    "\n",
+    "sub.index += 1\n",
+    "\n",
+    "\n",
+    "sub[['ID','Pred']].to_csv(\"ncaa-march-madness-submission.csv\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sub.count()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "_uuid": "ce959e232dd5c58f5748e74e7a5fe0c5d2d6ef6f"
+   },
+   "outputs": [],
+   "source": [
+    "from .generate_bracket import build_bracket\n",
+    "\n",
+    "b = build_bracket(\n",
+    "    output_path=\"2021_bracket.png\",\n",
+    "    teamsPath=\"../../data/MTeams.csv\",\n",
+    "    seedsPath=\"../../data/MNCAATourneySeeds.csv\",\n",
+    "    slotsPath=\"../../data/MNCAATourneySlots.csv\",\n",
+    "    submissionPath=\"ncaa-march-madness-submission.csv\",\n",
+    "    year=2021\n",
+    ")"
```

Comment:
Sorry this is so gross, but the only issue I've had is being able to access the build_bracket function from generate_bracket.py. I figured making it common would be nice for trying out different models, but I keep getting modulenotfound errors. Maybe you can tell what I'm doing wrong.

Reply:
Jk I got it